### ISSUE TITLE
Hyperparameter optimization

### ISSUE THREAD

**Issue by jcohenadad** [ivadomed/model_seg_ms_mp2rage#58]
Related to #49 

To try:

- [x] 2D vs. 3D
- [x] DA
- [x] n_filter
- [x] Dropout
- [x] Loss
- [ ] Softseg
- [x] Cascaded model

**Comment by jcohenadad** [comment]
# Effect of DA

RandomReverse:

![image](https://user-images.githubusercontent.com/2482071/215002185-ff443141-f27e-4551-bba0-be91f4aec53e.png)

Affine:

![image](https://user-images.githubusercontent.com/2482071/215102639-17744b2e-93cf-4f53-bcea-2eb728b93bfb.png)

**Comment by jcohenadad** [comment]
# Effect of `n_filter`

n_filter 8 vs. 16: (16 is better)

![image](https://user-images.githubusercontent.com/2482071/214475839-9cf001bd-d0f7-44f4-820d-bad90fa775b1.png)

n_filter 16 vs. 32: (32 is more unstable)

<img width="406" alt="image" src="https://user-images.githubusercontent.com/2482071/214939479-c3197d4a-6903-474a-b237-93f33df0d813.png">

n_filter 16 vs. 12:

![image](https://user-images.githubusercontent.com/2482071/215101899-9c250a60-f6e6-43d5-bec7-94ea874e0db2.png)

**Comment by jcohenadad** [comment]
Dropout 0.3 vs. 0.2:

![image](https://user-images.githubusercontent.com/2482071/214475888-34d0f4e9-6d36-4750-b11d-a69f13284175.png)

**Comment by jcohenadad** [comment]
<details><summary>Softseg</summary>

```console
{
    "command": "train",
    "gpu_ids": [4],
    "path_output": "/home/GRAMES.POLYMTL.CA/p101317/data_nvme_p101317/model_seg_lesion_mp2rage_20230102_144854",
    "model_name": "model_seg_lesion_mp2rage",
    "debugging": true,
    "object_detection_params": {
        "object_detection_path": null,
        "safety_factor": [1.0, 1.0, 1.0]
    },
    "wandb": {
        "wandb_api_key": "9095e2bc9e4ab445d478c9c8a81759ae908be8c6",
        "project_name": "basel-mp2rage-lesion",
        "group_name": "3D-newGT-filt16-softseg",
        "run_name": "run-01",
        "log_grads_every": 100
    },
    "loader_parameters": {
        "path_data": ["/home/GRAMES.POLYMTL.CA/p101317/data_nvme_p101317/data_seg_mp2rage_20221220_124553/data_processed_lesionseg"],
        "subject_selection": {"n": [], "metadata": [], "value": []},
        "target_suffix": ["_lesion-manualHaris"],
        "extensions": [".nii.gz"],
        "roi_params": {
            "suffix": null,
            "slice_filter_roi": null
        },
	"contrast_params": {
            "training_validation": ["UNIT1"],
            "testing": ["UNIT1"],
            "balance": {}
        },
        "slice_filter_params": {
            "filter_empty_mask": false,
            "filter_empty_input": true
        },
        "slice_axis": "axial",
        "multichannel": false,
        "soft_gt": true,
        "bids_validate": true
    },
    "split_dataset": {
        "fname_split": null,
        "random_seed": 42,
        "split_method" : "participant_id",
        "data_testing": {"data_type": null, "data_value":[]},
        "balance": null,
        "train_fraction": 0.6,
        "test_fraction": 0.1
    },
    "training_parameters": {
        "batch_size": 16,
        "loss": {
            "name": "AdapWingLoss"
        },
        "training_time": {
            "num_epochs": 50,
            "early_stopping_patience": 50,
            "early_stopping_epsilon": 0.001
        },
        "scheduler": {
            "initial_lr": 0.00005,
            "lr_scheduler": {
                "name": "CosineAnnealingLR",
                "base_lr": 1e-5,
                "max_lr": 1e-3
            }
        },
        "balance_samples": {
            "applied": false,
            "type": "gt"
        },
        "mixup_alpha": null,
        "transfer_learning": {
            "retrain_model": null,
            "retrain_fraction": 1.0,
            "reset": true
        }
    },
    "default_model": {
        "name": "Unet",
        "dropout_rate": 0.2,
        "bn_momentum": 0.1,
        "final_activation": "relu",
        "depth": 3
    },
    "FiLMedUnet": {
        "applied": false,
        "metadata": "contrasts",
        "film_layers": [0, 1, 0, 0, 0, 0, 0, 0, 0, 0]
    },
    "Modified3DUNet": {
        "applied": true,
        "length_3D": [32, 64, 64],
        "stride_3D": [1, 1, 64],
        "attention": false,
        "n_filters": 16
    },
    "uncertainty": {
        "epistemic": false,
        "aleatoric": false,
        "n_it": 0
    },
    "postprocessing": {
        "remove_noise": {"thr": 0.01},
        "remove_small": {"unit": "vox", "thr": 3}
    },
    "evaluation_parameters": {
        "target_size": {"unit": "vox", "thr": [20, 100]},
        "overlap": {"unit": "vox", "thr": 3}
    },
    "transformation": {
        "Resample": {"hspace": 1, "wspace": 1, "dspace": 1},
        "CenterCrop": {"size": [32, 64, 128]},
	"NormalizeInstance": {"applied_to": ["im"]}
    }
}
```

</details>

![image](https://user-images.githubusercontent.com/2482071/214699864-4a04f795-3556-47d6-888e-f1dc6a9e0736.png)

Same config, with omega=8 (instead of 14), as per SoftSeg MS lesion paper (not better):

![image](https://user-images.githubusercontent.com/2482071/214715764-0b499530-2fc8-40c4-bc3b-00a9551a3175.png)

> **Warning**
> The test above was using Haris' ground truth

**Comment by jcohenadad** [comment]
<details><summary>Cascaded model</summary>

```json
{
    "command": "train",
    "gpu_ids": [4],
    "path_output": "/home/GRAMES.POLYMTL.CA/p101317/data_nvme_p101317/model_seg_lesion_mp2rage_20230102_144854",
    "model_name": "model_seg_lesion_mp2rage",
    "debugging": true,
    "log_file": "train.log",
    "wandb": {
        "wandb_api_key": "9095e2bc9e4ab445d478c9c8a81759ae908be8c6",
        "project_name": "basel-mp2rage-lesion",
        "group_name": "3D-newGT-filt16-cascade",
        "run_name": "run-01",
        "log_grads_every": 100
    },
    "loader_parameters": {
        "path_data": ["/home/GRAMES.POLYMTL.CA/p101317/data_nvme_p101317/data.neuro/basel-mp2rage"],
        "subject_selection": {"n": [], "metadata": [], "value": []},
        "target_suffix": ["_lesion-manualNeuroPoly"],
        "extensions": [".nii.gz"],
        "roi_params": {
            "suffix": null,
            "slice_filter_roi": null
        },
	"contrast_params": {
            "training_validation": ["UNIT1"],
            "testing": ["UNIT1"],
            "balance": {}
        },
        "slice_filter_params": {
            "filter_empty_mask": false,
            "filter_empty_input": true
        },
        "slice_axis": "axial",
        "multichannel": false,
        "soft_gt": false,
        "bids_validate": true
    },
    "split_dataset": {
        "fname_split": null,
        "random_seed": 42,
        "split_method" : "participant_id",
        "data_testing": {"data_type": null, "data_value":[]},
        "balance": null,
        "train_fraction": 0.7,
        "test_fraction": 0.1
    },
    "object_detection_params": {
        "object_detection_path": "/home/GRAMES.POLYMTL.CA/p101317/data_nvme_p101317/model_seg_ms_sc_mp2rage",
        "safety_factor": [1.0, 1.0, 1.0]
    },
    "training_parameters": {
        "batch_size": 16,
        "loss": {
            "name": "DiceLoss"
        },
        "training_time": {
            "num_epochs": 100,
            "early_stopping_patience": 50,
            "early_stopping_epsilon": 0.001
        },
        "scheduler": {
            "initial_lr": 0.001,
            "lr_scheduler": {
                "name": "CosineAnnealingLR",
                "base_lr": 1e-5,
                "max_lr": 1e-3
            }
        },
        "balance_samples": {
            "applied": false,
            "type": "gt"
        },
        "mixup_alpha": null,
        "transfer_learning": {
            "retrain_model": null,
            "retrain_fraction": 1.0,
            "reset": true
        }
    },
    "default_model": {
        "name": "Unet",
        "dropout_rate": 0.3,
        "bn_momentum": 0.1,
        "final_activation": "sigmoid",
        "depth": 3
    },
    "FiLMedUnet": {
        "applied": false,
        "metadata": "contrasts",
        "film_layers": [0, 1, 0, 0, 0, 0, 0, 0, 0, 0]
    },
    "Modified3DUNet": {
        "applied": true,
        "length_3D": [32, 64, 64],
        "stride_3D": [1, 1, 64],
        "attention": false,
        "n_filters": 16
    },
    "uncertainty": {
        "epistemic": false,
        "aleatoric": false,
        "n_it": 0
    },
    "postprocessing": {
        "remove_noise": {"thr": 0.01},
        "binarize_prediction": {"thr": 0.5},
        "remove_small": {"unit": "vox", "thr": 3}
    },
    "evaluation_parameters": {
        "target_size": {"unit": "vox", "thr": [20, 100]},
        "overlap": {"unit": "vox", "thr": 3}
    },
    "transformation": {
        "Resample": {"hspace": 1, "wspace": 1, "dspace": 1},
        "CenterCrop": {"size": [32, 64, 128]},
	"NormalizeInstance": {"applied_to": ["im"]}
    }
}
```

</details>

![image](https://user-images.githubusercontent.com/2482071/214880474-5f8d6da5-6b9d-4a82-aafe-e4e8f6af1531.png)

> **Note**
> Training is _very_ slow due to the application of the model on each input data

**Comment by jcohenadad** [comment]
Best loss: DiceLoss

![image](https://user-images.githubusercontent.com/2482071/214966605-d7bd1e91-4dcf-4004-8325-182dc2a1e8fe.png)

**Comment by jcohenadad** [comment]
3D better than 2D:

![image](https://user-images.githubusercontent.com/2482071/215002097-40096624-c2b9-429e-a6df-2de0c1bb054b.png)

<details><summary>config 2D model</summary>

```console
{
    "command": "train",
    "gpu_ids": [4],
    "path_output": "/home/GRAMES.POLYMTL.CA/p101317/data_nvme_p101317/model_seg_lesion_mp2rage_20230102_144854",
    "model_name": "model_seg_lesion_mp2rage",
    "debugging": true,
    "log_file": "train.log",
    "wandb": {
        "wandb_api_key": "9095e2bc9e4ab445d478c9c8a81759ae908be8c6",
        "project_name": "basel-mp2rage-lesion",
        "group_name": "2D-newGT",
        "run_name": "run-01",
        "log_grads_every": 100
    },
    "loader_parameters": {
        "path_data": ["/home/GRAMES.POLYMTL.CA/p101317/data_nvme_p101317/data_seg_mp2rage_20230124_132415/data_processed_lesionseg"],
        "subject_selection": {"n": [], "metadata": [], "value": []},
        "target_suffix": ["_lesion-manualNeuroPoly"],
        "extensions": [".nii.gz"],
        "roi_params": {
            "suffix": null,
            "slice_filter_roi": null
        },
	"contrast_params": {
            "training_validation": ["UNIT1"],
            "testing": ["UNIT1"],
            "balance": {}
        },
        "slice_filter_params": {
            "filter_empty_mask": false,
            "filter_empty_input": true
        },
        "slice_axis": "axial",
        "multichannel": false,
        "soft_gt": false,
        "bids_validate": true
    },
    "split_dataset": {
        "fname_split": null,
        "random_seed": 42,
        "split_method" : "participant_id",
        "data_testing": {"data_type": null, "data_value":[]},
        "balance": null,
        "train_fraction": 0.7,
        "test_fraction": 0.1
    },
    "object_detection_params": {
        "object_detection_path": null,
        "safety_factor": [1.0, 1.0, 1.0]
    },
    "training_parameters": {
        "batch_size": 16,
        "loss": {
            "name": "DiceLoss"
        },
        "training_time": {
            "num_epochs": 50,
            "early_stopping_patience": 50,
            "early_stopping_epsilon": 0.001
        },
        "scheduler": {
            "initial_lr": 0.001,
            "lr_scheduler": {
                "name": "CosineAnnealingLR",
                "base_lr": 1e-5,
                "max_lr": 1e-3
            }
        },
        "balance_samples": {
            "applied": false,
            "type": "gt"
        },
        "mixup_alpha": null,
        "transfer_learning": {
            "retrain_model": null,
            "retrain_fraction": 1.0,
            "reset": true
        }
    },
    "default_model": {
        "name": "Unet",
        "dropout_rate": 0.3,
        "bn_momentum": 0.1,
        "final_activation": "sigmoid",
        "depth": 3
    },
    "FiLMedUnet": {
        "applied": false,
        "metadata": "contrasts",
        "film_layers": [0, 1, 0, 0, 0, 0, 0, 0, 0, 0]
    },
    "Modified3DUNet": {
        "applied": false,
        "length_3D": [32, 64, 64],
        "stride_3D": [1, 1, 64],
        "attention": false,
        "n_filters": 16
    },
    "uncertainty": {
        "epistemic": false,
        "aleatoric": false,
        "n_it": 0
    },
    "postprocessing": {
        "remove_noise": {"thr": 0.01},
        "binarize_prediction": {"thr": 0.5},
        "remove_small": {"unit": "vox", "thr": 3}
    },
    "evaluation_parameters": {
        "target_size": {"unit": "vox", "thr": [20, 100]},
        "overlap": {"unit": "vox", "thr": 3}
    },
    "transformation": {
        "Resample": {"hspace": 1, "wspace": 1, "dspace": 1},
        "CenterCrop": {"size": [32, 64]},
	"NormalizeInstance": {"applied_to": ["im"]}
    }
}
(base) p101317@ro
```

</details>

**Comment by jcohenadad** [comment]
Good model was found
